lib/elixir/lib/kernel.ex

```diff
@@ -957,7 +971,10 @@ defmodule Kernel do
       [1, 2 | 3]
 
   """
-  @spec (list ++ term) :: maybe_improper_list
+  @spec (nonempty_list ++ nonempty_list) :: nonempty_list
+  @spec (left ++ []) :: left when left: nonempty_list
+  @spec ([] ++ right) :: right when right: term
+  @spec (list ++ nonempty_improper_list(any, any)) :: nonempty_improper_list(any(), any())
```

Comment:
Please let's keep the original one. Again, this exists mostly as a documentation purpose, and I think the current one is the clearest documentation wise.